### ISSUE TITLE
custom/config doesn't override config

### ISSUE THREAD

**Issue by adi8i** [gantry/gantry5#1346]
WordPress
Version: dev-71b8b9d06 / Released: June 7, 2016

I did an outline that's created by default, `config/home...` it have atoms too. Now if I want to remove the atoms from home outline and inherit them from base it's not possible because the atoms will be re added by `config/home/page/head.yaml`, the only way to remove them is to save the section empty.

**Comment by w00fz** [comment]
I am not quite sure about this one @adi8i.
Are you trying to inherit from Base or are you trying to have the Atoms section empty?

**Comment by adi8i** [comment]
Hey. I was trying to inherit from base, but was not possible because of `config/home/page/head.yaml`, that keep re-adding the atoms.

https://drive.google.com/file/d/0B4NiUo_62I-AUl9ObnRqaE8yaWs/view?usp=sharing

**Comment by mahagr** [comment]
Honestly we are moving away from having pre-configured stuff in `config` folder, mostly because of its not working in Joomla (which uses dynamic ids) and because of some merging (data coming back etc) issues on the configuration itself.

**Comment by adi8i** [comment]
You're right, it keep re-adding things back, including styles. It was a good way to install demo data.

**Comment by JoomFX** [comment]
@mahagr Yes, Adi is right.
I mean, how would you pass the default template configuration if you remove the `config` folder (speaking about the one that is outside of the `custom` folder)?

You (RT) and I build all our templates using the `config` folder, that's how if the user installs just the template he will get all the settings and all the necessary atoms loaded by default.

How would we do that without the `config` folder, or am I missing something here?

**Comment by mahagr** [comment]
Preferred way in Joomla is to install config by adding files into `install/layouts` which allows installer to properly deal with renaming outlines during installation.

I'm going to keep this issue open, though, as I know we need to improve installation process a bit.

**Comment by JoomFX** [comment]
Hmm, alright thanks for the info Matias.

I've never had any issues with the `config` folder.
I also just had a look at Topaz and it seems that you use `install/layouts` for the "Page Settings" only. The rest is still being added by the `config` folder.

I would love if we can keep the `config` approch... to improve it if really needed. Once again, none of our customers have ever reported any installation issues or issues related to it :wink:

P.S. Probably I have not experience any issues, because in the `config` folder we have stuff for the "Default" (and "Error") outline only and this outline inherits everything (Styles and Atoms) from "Base".

**Comment by adi8i** [comment]
The issue is on WordPress ;)

**Comment by JoomFX** [comment]
Thanks Adi :smile: 

Yeah, I got confused because Matias speaks about issues in Joomla...

**Comment by adi8i** [comment]
@mahagr @w00fz I tested more this issue, i'll try to explain what's going on. 

In my case is an image field, if you save the field empty, in configuration it will be saved as `""` and that's fine, but the checkbox doesn't stay checked, now if you refresh the page and hit save again, or you need to do some changes at a later time to other fields, because the field is no more checked it will be removed from configurations and the value will inherit from `base` or `config`.

Every time you need to adjust `styles` you also need to save the image field empty.

https://drive.google.com/file/d/0B4NiUo_62I-AdTFSODVkclZKbHc/view

**Comment by w00fz** [comment]
But what is the default value in Base Outline Styles? Is it empty? If it's empty this is the proper behavior. The check shows checked only if it's different.

**Comment by adi8i** [comment]
The value in base outline is empty. I understand, but this way if you have a anything in `config` folder you can't save the field empty, well you can but you need to save it empty each time you adjust styles.

**Comment by w00fz** [comment]
If your default in Base was stored as `bla`, then you can save empty and would stay that way, otherwise the way Gantry work is to just use Base if the value match. It's identical to the colors, empty is just another value

**Comment by adi8i** [comment]
It use `template_name/config/test/styles.yaml`. True, if the default is `bla` you can save it empty. But in same circumstances it's annoying. 

For example i use `template_name/config/...` to install demo outlines on WordPress. In base outline all image fields are empty, because it is used as fallback for wp things or for posts. Test outline have bg images, now if the user want to remove those images... 

I guess you install outlines in the same way. Would be better to use a function to create them, similar to Joomla, @newkind what do you think ?

**Comment by mahagr** [comment]
Related to #1367, even if that one is for the menu. Basically we need to change how saving works so it stores all the altered values, not just the changed ones.

**Comment by mahagr** [comment]
Alright, I've implemented persistent configuration overrides. So basically if you set override value, it will stay until you uncheck it.

**Comment by adi8i** [comment]
Ehe, you changed few more things. Thanks a lot Matias !

**Comment by mahagr** [comment]
I'm not sure if it fixes all the issues with config vs custom/config, but at least it should allow a workaround.

**Comment by JoomFX** [comment]
Thanks Matias!
So we can still use the `config` folder approach and be save? I mean there's no need to use the `install/layouts`, right?

**Comment by mahagr** [comment]
@JoomFX I'm not sure if there are some other issues as well; I believe that there may be some. This is surely something I need to take a closer look and find proper solution for all the platforms.

**Comment by adi8i** [comment]
@mahagr can it be merged in development ?

**Comment by mahagr** [comment]
Merged.

**Comment by w00fz** [comment]
This enhancement introduced a regression which I now fixed. Please make sure the persistent override still works.

/ref. 27e7c5d

**Comment by adi8i** [comment]
@mahagr the array thing is still there

**Comment by mahagr** [comment]
It was the fix @w00fz reverted. I need to find out another way to fix it without breaking defaults.

**Comment by w00fz** [comment]
@mahagr this seems to be still an issue with the `config` vs `custom/config`. 

As per @adi8i https://github.com/gantry/gantry5/issues/1402#issuecomment-231485780 
> This is still an issue + #1346 
>
> When i say config i mean `template-name/config` 
> 
> In this example the field is checked by default, if i save it empty it should inherit from base. But nothing happens. After refresh, the field is checked again.
> http://www.awesomescreenshot.com/image/1383069/13baa651829e689abfafcc9d4ca5b125
> 
> I think you understood it wrong. The default behavior is fine, empty field -> inherit from base.
> 
> What i mean about config, is that it should be used to install demo data only, to create outlines on install. Like you do with home outline in hydrogen, wordpress. Or i don't understand how config works :P. Please let's me know if i do something wrong.
> 
> There are two states. Field unchecked -> inherit from base. Field checked -> override base. Config likely introduce a third state and that's not possible.

**Comment by adi8i** [comment]
I tried now in hydrogen. What happens is in a way normal thought. If you have `g5_hydrogen/config/home/styles.yaml` when a field is unchecked it inherit from config.

**Comment by mahagr** [comment]
That's expected behavior. Config has a value, so it overrides base. This is why we're now trying to get rid of config folder and have install folder instead, which gets copied during installation. I'll look into this and will get it fully working for G5.3.1...

**Comment by adi8i** [comment]
Yep, thanks Matias, that i was thinking last night, but as i was too tired, I was writing bs
